app/data/serializers.py

```diff
@@ -48,6 +53,14 @@ def filter_details_only(self, key, value):
             raise serializer_fields.DropJsonKeyException
 
 
+class DetailsReadOnlyRecordNonPublicSerializer(DetailsReadOnlyRecordSerializer):
+    """
+    Serialize records with only read-only fields included plus non-public fields
+    (only available to admins and analysts)
+    """
+    created_by = CharField()
```

Comment:
I'm a bit worried that having multiple serializers for the same model and view is a bit of an antipatern. Some day I'd like to see what it'd take to consolidate our ReadOnly/Write sererializers, but for this [from reading online](http://masnun.com/2015/10/21/django-rest-framework-dynamic-fields-in-serializers.html) I think we can add the field dynamically in `__init__()`:
```
def __init__(self, *args, **kwargs):
    super(OrderSerializer, self).__init__(*args, **kwargs)
 
    if is_admin_or_writer(self.request.user):
        self.fields['created_by']  = CharField()
```

Reply:
I'm having trouble getting this working (`self.context` is empty). I'm also not entirely sure I like the idea of having fewer serializers whose fields we change dynamically. Seems like it might be hard to understand and maintain.

Reply:
Sorry, I didn't realize that VSCode publishes PR comments immediately and edited my comment after I posted. Tested afterwards and `self.context['request']` should work.

I may be overly cautious from getting burned by creating subclasses and discovering months later that their pivot factors weren't as fundamental as I originally though. My concern is that class inheritance doesn't scale the way we'd likely need for controlling the fields in a given response. Inheritance forces a linear decision tree, where the process of selecting a serializer class must follow down one pre-determined branch first, then the next, etc. It works well if each decision is a categorical distinction (Like a JSON parser class vs CSV parser class), but as the decisions become less categorical and more business logic it raises the risk that you have conflicting decision points.

Like if we want to add another level of controlling a field based on a factor orthogonal to user role (Such as a feature flag, perhaps), it isn't clear how the levels would be structured - worst case, if they are truly independent we'd need a cartesian product of serializer classes. In that case, although it's decidedly less elegant, I think having fields dynamically determined in code would be the only maintainable solution.

Reply:
All that said, it's not like doing class inheritance or dynamic fields for this PR specifically is an inflection point that we can't revisit later on. It's pretty squarely a Type 2 decision, so if we like class inheritance more for now then 👍

Reply:
Late to this party but one thing I've noticed is that DRF sometimes gets in the way of making serializers too intelligent, so the "different classes plus switching logic in `get_serializer_class`" is often a good way to make DRF happy even if it sometimes feels a bit strange to have a profusion of serializer subclasses.